oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/linkedList/PersistedLinkedListV2.java

```diff
@@ -0,0 +1,293 @@
+    private long peakCacheSizeBytes;
+    private long peakCacheSize;
+
+    /**
+     * @param cacheSize   the maximum number of elements to keep in the in-memory cache
+     * @param cacheSizeMB the maximum size of the in-memory cache in MB
+     */
+    public PersistedLinkedListV2(String fileName, NodeStateEntryWriter writer, NodeStateEntryReader reader, int cacheSize, int cacheSizeMB) {
+        this.cacheSizeLimit = cacheSize;
+        this.cacheSizeLimitBytes = ((long) cacheSizeMB) * 1024 * 1024;
+        this.storeFileName = fileName;
+        LOG.info("Opening store {}", fileName);
+        File oldFile = new File(fileName);
+        if (oldFile.exists()) {
+            LOG.info("Deleting {}", fileName);
+            try {
+                FileUtils.forceDelete(oldFile);
+            } catch (IOException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+        openStore();
+        this.writer = writer;
+        this.reader = reader;
+        lastCompact = System.currentTimeMillis();
+    }
+
+    private void openStore() {
+        store = MVStore.open(storeFileName);
+        map = store.openMap("list");
+    }
+
+    @Override
+    public void add(@NotNull NodeStateEntry item) {
+        Preconditions.checkArgument(item != null, "Can't add null to the list");
+        Long index = tailIndex++;
+        addEntryToCache(index, item);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return totalEntries == 0;
+    }
+
+    @Override
+    public Iterator<NodeStateEntry> iterator() {
+        return new NodeIterator(headIndex);
+    }
+
+    @Override
+    public NodeStateEntry remove() {
+        Preconditions.checkState(!isEmpty(), "Cannot remove item from empty list");
+        Long boxedHeadIndex = headIndex;
+        NodeStateEntry entryRemoved = cache.remove(boxedHeadIndex);
+        if (entryRemoved == null) {
+            String mapEntry = map.remove(boxedHeadIndex);
+            if (mapEntry == null) {
+                throw new IllegalStateException("Entry not found in cache or in store: " + boxedHeadIndex);
+            }
+            cacheMisses++;
+            entryRemoved = reader.read(mapEntry);
+        } else {
+            cacheHits++;
+            cacheSizeEstimationBytes -= entryRemoved.estimatedMemUsage();
+        }
+
+        headIndex++;
+        totalEntries--;
+        if (totalEntries == 0) {
+            map.clear();
+            cache.clear();
+        }
+        return entryRemoved;
+    }
+
+    private NodeStateEntry get(Long index) {
+        NodeStateEntry result = cache.get(index);
+        if (result == null) {
+            cacheMisses++;
+            String s = map.get(index);
+            result = reader.read(s);
+            LOG.trace("Cache miss: {}={}", index, result.getPath());
+        } else {
+            cacheHits++;
+        }
+        return result;
+    }
+
+    private void addEntryToCache(Long index, NodeStateEntry entry) {
```

Comment:
I would use "long" instead of "Long" here, because index may not be null. (I'm aware it can be null in the internal hash map.)

Reply:
Done. I initially chose Long in an earlier version of this method to reduce boxing. But in the current version, it already does only one boxing in all possible execution paths, so indeed this optimization is no longer necessary.